lib/Driver/Driver.cpp

```diff
@@ -643,6 +643,88 @@ static bool getFilelistThreshold(DerivedArgList &Args, size_t &FilelistThreshold
   return false;
 }
 
+void Driver::setBatchSeed(const llvm::opt::InputArgList *ArgList) {
+  if (const Arg *A = ArgList->getLastArg(options::OPT_driver_batch_seed)) {
+    if (StringRef(A->getValue()).getAsInteger(10, DriverBatchSeed)) {
+      Diags.diagnose(SourceLoc(), diag::error_invalid_arg_value,
+                     A->getAsString(*ArgList), A->getValue());
+    }
+  }
+}
+
+static bool computeIncremental(const llvm::opt::InputArgList *ArgList,
+                               const bool ShowIncrementalBuildDecisions) {
+  const bool WasIncrementalRequested =
+      ArgList->hasArg(options::OPT_incremental);
+  const char *ReasonToDisable =
+      ArgList->hasArg(options::OPT_whole_module_optimization)
+          ? "is not compatible with whole module optimization."
+          : ArgList->hasArg(options::OPT_embed_bitcode)
+                ? "is not currently compatible with embedding LLVM IR bitcode"
+                : nullptr;
+
+  if (!WasIncrementalRequested || !ReasonToDisable)
+    return WasIncrementalRequested;
+
+  if (ShowIncrementalBuildDecisions) {
+    llvm::outs() << "Incremental compilation has been disabled, because it "
+                 << ReasonToDisable;
+  }
+  return false;
+}
+
+static std::string
+computeWorkingDirectory(const llvm::opt::InputArgList *ArgList) {
+  if (auto *A = ArgList->getLastArg(options::OPT_working_directory)) {
+    SmallString<128> workingDirectory;
+    workingDirectory = A->getValue();
+    llvm::sys::fs::make_absolute(workingDirectory);
+    return workingDirectory.str();
```

Comment:
Thank you for this. Yes. In this case, the copying overhead is not significant, we get to shrink down a super-long function, and encapsulate the intension-shift.

Reply:
I think Robert was saying that the implicit conversion here from StringRef to std::string was too subtle for him. Is that correct?

Reply:
I'm not sure what "implicit" means here because I had thought that the `.str()` part as well as the return type made the conversion explicit. I would like to encapsulate this computation in a function in order to capture the intension-shift (aka abstract-level change), from the caller to this callee. Is there a way to do it that you would prefer? How else would one write a function that returns a string?

Reply:
I'll let Robert decide what he'd prefer—I have no further comments—but `str()` produces a StringRef, not a std::string. The maximally explicit thing would be \*sigh\* `workingDirectory.str().str()`.

Reply:
@jrose-apple Thanks! I didn't realize that--not familiar enough with SmallString and `str` didn't convey that to me. I have changed that to make it clearer.
@CodaFi How do you feel about how it reads now?